phoenix-core/src/it/java/org/apache/phoenix/jdbc/HighAvailabilityGroupIT.java

```diff
@@ -0,0 +1,675 @@
+        clientProperties.setProperty(PHOENIX_HA_GROUP_ATTR, haGroupName2);
+
+        // create cluster role records with different versions on two ZK clusters
+        final String zpath = ZKPaths.PATH_SEPARATOR + haGroupName2;
+        ClusterRoleRecord record1 = new ClusterRoleRecord(
+                haGroupName2, HighAvailabilityPolicy.FAILOVER,
+                CLUSTERS.getUrl1(), ClusterRole.ACTIVE,
+                CLUSTERS.getUrl2(), ClusterRole.STANDBY,
+                1);
+        CLUSTERS.createCurator1().create().forPath(zpath, ClusterRoleRecord.toJson(record1));
+        ClusterRoleRecord record2 = new ClusterRoleRecord(
+                record1.getHaGroupName(), record1.getPolicy(),
+                record1.getZk1(), record1.getRole1(),
+                record1.getZk2(), record1.getRole2(),
+                record1.getVersion() + 1); // record2 is newer
+        CLUSTERS.createCurator2().create().forPath(zpath, ClusterRoleRecord.toJson(record2));
+
+        doTestWhenOneZKDown(CLUSTERS.getHBaseCluster2(), () -> {
+            Optional<HighAvailabilityGroup> haGroup2 = HighAvailabilityGroup.get(jdbcUrl, clientProperties);
+            assertTrue(haGroup2.isPresent());
+            assertNotSame(haGroup2.get(), haGroup);
+            // apparently the HA group cluster role record should be from the healthy cluster
+            assertEquals(record1, haGroup2.get().getRoleRecord());
+        });
+
+        // When ZK2 is connected, its cluster role manager should apply newer cluster role record
+        waitFor(() -> {
+            try {
+                Optional<HighAvailabilityGroup> haGroup2 = HighAvailabilityGroup.get(jdbcUrl, clientProperties);
+                return haGroup2.isPresent() && record2.equals(haGroup2.get().getRoleRecord());
+            } catch (SQLException e) {
+                LOG.warn("Fail to get HA group {}", haGroupName2);
+                return false;
+            }
+        }, 100, 30_000);
+
+        // clean up HA group 2
+        HighAvailabilityGroup.get(jdbcUrl, clientProperties).ifPresent(HighAvailabilityGroup::close);
+    }
+
+    /**
+     * Test that HA connection request will fall back to the first cluster when HA group fails
+     * to initialize due to missing cluster role record (CRR).
+     */
+    @Test
+    public void testFallbackToSingleConnection() throws Exception {
+        final String tableName = testName.getMethodName();
+        CLUSTERS.createTableOnClusterPair(tableName);
+
+        String haGroupName2 = testName.getMethodName() + RandomStringUtils.randomAlphabetic(3);
+        clientProperties.setProperty(PHOENIX_HA_GROUP_ATTR, haGroupName2);
+        // no cluster role record for the HA group with name haGroupName2
+        try (Connection conn = DriverManager.getConnection(jdbcUrl, clientProperties)) {
+            // connection is PhoenixConnection instead of HA connection (failover or parallel)
```

Comment:
Interesting (and good) fallback behavior. I'm sure it's somewhere below, but do we log anywhere that this fallback happened? (Seems like a WARN level sort of thing.)

Reply:
We do log in the classes under test.